src/common/adapter_manager.rs

```diff
@@ -47,50 +52,59 @@ where
     pub fn emit(&self, event: CentralEvent) {
         match event {
             CentralEvent::DeviceDisconnected(addr) => {
-                self.peripherals.remove(&addr);
+                self.shared.peripherals.remove(&addr);
             }
             CentralEvent::DeviceLost(addr) => {
-                self.peripherals.remove(&addr);
+                self.shared.peripherals.remove(&addr);
             }
             _ => {}
         }
 
-        send_notification(&self.async_senders, &event);
+        // Note: we ignore possible send errors which are expected while there
+        // could be zero recievers if application is not subscribed
+        let _ = self.shared.events_channel.send(event);
     }
 
     pub fn event_stream(&self) -> Pin<Box<dyn Stream<Item = CentralEvent> + Send>> {
-        let (sender, receiver) = mpsc::unbounded();
-        self.async_senders.lock().unwrap().push(sender);
-        Box::pin(receiver)
+        let receiver = self.shared.events_channel.subscribe();
+        Box::pin(BroadcastStream::new(receiver).filter_map(|x| async move {
+            if x.is_ok() {
+                Some(x.unwrap())
+            } else {
+                None
+            }
+        }))
```

Comment:
You have the same pattern somewhere else in a previous PR, could you factor it out to a helper function?

Reply:
Although it's equivalent to what's done for the characteristic notification events I wonder if it's worth having a utility while it's semantically a different event stream (theoretically it seems plausible that it could make sense to do different things here than is done for a characteristic notification stream)?

There's a repetition of the same thing between the winrt backend and corebluetooth backend too and I suppose a utility could be added for those, which could probably technically be used here too, but maybe it'd be more worthwhile having a wrapper if it's also for a more specific purpose like for a notification stream?

Maybe I could add a utility at the same time as updating the corebluetooth backend - just for notifications to start with and also remove `send_notification` which is now not used by anything?

Reply:
Okey, for reference here I've updated my PR that updates the corebluetooth backend to add a utility for this: https://github.com/deviceplug/btleplug/pull/187

https://github.com/deviceplug/btleplug/pull/187/commits/76c0fc39c1ccfacd4b4f54bc1033e5f3402ba9fa